### PR TITLE
Refactor fallback code to prepare for never type

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This PR contains cherry-picks of some of @nikomatsakis's work from #79366, and shouldn't (AFAICT) represent any change in behavior. However, the refactoring is good regardless of the never type work being landed, and will reduce the size of those eventual PR(s) (and rebase pain).

I am not personally an expert on this code, and the commits are essentially 100% @nikomatsakis's, but they do seem reasonable to me by my understanding. Happy to edit with review, of course. Commits are best reviewed in sequence rather than all together.

r? @jackh726 perhaps?

### CODE REVIEW

File: compiler/rustc_typeck/src/check/fallback.rs [hunk 1]
```diff
@@ -43,10 +45,113 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
...
+        // resolved to a special integral type, `?T` is not considered unsolved,
+        // but `?I` is. The same is true for float variables.)
+        let fallback = match ty.kind() {
+            _ if self.is_tainted_by_errors() => self.tcx.ty_error(),
+            ty::Infer(ty::IntVar(_)) => self.tcx.types.i32,
+            ty::Infer(ty::FloatVar(_)) => self.tcx.types.f64,
+            _ => match self.type_var_diverges(ty) {
+                Diverging::Diverges => self.tcx.mk_diverging_default(),
+                Diverging::NotDiverging => return false,
+            },
```
Comment (jackh726) [inline compiler/rustc_typeck/src/check/fallback.rs:91]: Oh this looks much nicer now.

**Review by jackh726 — COMMENTED** [review]
Overall, the changes here are pretty simple. Have a few comments/thoughts, but I'd like to take another pass before r+ing.

File: compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs [hunk 2]
```diff
@@ -656,56 +657,60 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                // We treat this as a non-defining use by making the inference
-                // variable fall back to the opaque type itself.
-                if let FallbackMode::All = mode {
-                    if let Some(opaque_ty) = self.infcx.inner.borrow().opaque_types_vars.get(ty) {
-                        debug!(
-                            "fallback_if_possible: falling back opaque type var {:?} to {:?}",
-                            ty, opaque_ty
-                        );
-                        *opaque_ty
+            Neither => match self.type_var_diverges(ty) {
```
[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs]: This is an intermediate commit (and now the contents have moved to fallback.rs) but I'm not sure I follow -- there's an  early return in the notdiverging case, so it seems a little painful to do those?

[the remark this answers is not in the document]
Comment (jackh726) [reply compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs]: Yep. This is better later.

**Comment by Mark-Simulacrum** [comment]
Force-pushed squashed in fixes to your current comments (since they were all mostly "nit"-like).

> As I read through the commits, I'm curious how this will get used.

https://gist.github.com/nikomatsakis/7a07b265dc12f5c3b3bd0422018fa660 is a pretty good write up of the "v1" algorithm proposed for landing never type. There's some additional extensions that are currently suggested in #84573 to mitigate some of the cases v1 doesn't address, but those are about detecting particular edge cases really.

The goal of the coercion predicates in particular is to build up the coercion graph added in the commits after these ~8 in #84573, which is an attempt to track the "liveness" of particular type variables during fallback.

> I'm curious what the diagnostics look like without this.

You can see a diff here - https://github.com/rust-lang/rust/commit/acd7444b2ace8e1873dcb52101c535cc75ee8cbe -- nothing obviously major either way, though issue-57843 under NLL interestingly compiles without errors?

**Comment by jackh726** [comment]
Yes, I started to go back and read through the write up.

It's really interesting  that issue-57843 compiles in nll without that extra sub. (Actually, I feel like the behavior with the change is more of what we *want*, but it's still a change). This makes me a bit suspicious. At the very least the comments need to be changed to address that it's not purely diagnostics. Maybe @nikomatsakis might be interested or have thoughts.

(P.s. I think it's funny and a nice coincidence how similar 84573 and 57843 are to each other)

**Comment by jackh726** [comment]
@Mark-Simulacrum 536da6f679cae3a1559f9a368e08fdf97c9c6148 can be removed in isolation, right? If you can remove that, then r=me on this. Given that it does *not* end up as purely diagnostics changes, I'm more wary to approve that right now. Given that everything else is straightforward, I wouldn't one that one commit to up the whole PR.

So, r=me on the other 7 commits.

**Comment by jackh726** [comment]
Ah interesting. Seems less straightforward than I thought. In this case, I'm fine with adding the commit back in with a proper comment that the extra sub is not only for diagnostics but may have semantic effects.

**Comment by Mark-Simulacrum** [comment]
Added the note here https://github.com/rust-lang/rust/pull/88149/commits/020655b90da83b0a037a9d3c987be17f04eedbc4#diff-79cce3ca916fa3f7d4f1b1879dbec1b3852414a5cfc719087851635aaa240a97R1018-R1019, will see if CI is passing (just in case, though it really should).

**Pull request merged**